py/packages/genkit/src/genkit/core/aio.py

```diff
@@ -1,109 +1,194 @@
+                [pop_task, self._close_future],
+                return_when=asyncio.FIRST_COMPLETED,
+                timeout=self._timeout,
+            )
+
+        except TimeoutError as e:
+            # If the wait operation timed out, both tasks will be in pending.
+            # asyncio.wait() doesn't automatically cancel pending tasks, so we
+            # ensure we cancel our pending tasks.
+            if not finished and self._timeout is not None:
+                # This means we've timed out, so cancel the pending tasks.
+                for t in pending:
+                    t.cancel()
+                raise e
+
+        # If the pop task completed, return its result.
+        if pop_task in finished:
+            return pop_task.result()
+
+        # If the close future completed, raise StopAsyncIteration.
+        if self._close_future in finished:
+            # Cancel pop task if we're done, avoid warnings.
+            pop_task.cancel()
+            raise StopAsyncIteration
+
+        # Wait for the pop task with a timeout, raise TimeoutError if a timeout
+        # is specified and is exceeded and automatically cancel the pending
+        # task.
+        return await asyncio.wait_for(pop_task, timeout=self._timeout)
+
+    def send(self, value: T) -> None:
+        """Sends a value into the channel.
+
+        The value is added to the internal queue for consumers to retrieve.
+        This is a non-blocking operation.
 
         Args:
-            value: The value to send.
-        """
-        return self.queue.put_nowait(value)
+            value: The value to send through the channel.
+
+        Raises:
+            asyncio.QueueFull: If the channel's internal queue is full.
 
-    def set_close_future(self, future: Future):
+        Returns:
+            None.
         """
-        Sets a future that, when completed, will close the channel.
+        self.queue.put_nowait(value)
+
+    def set_close_future(self, future: asyncio.Future) -> None:
+        """Sets a future that, when completed, will close the channel.
+
+        When the provided future completes, the channel will be marked as
+        closed, signaling to consumers that no more values will be sent.
 
         Args:
-            f (Future): The future to set.
-        """
-        self.__close_future = ensure_future(future)
-        self.__close_future.add_done_callback(
-            lambda v: self.closed.set_result(v.result())
-        )
+            future: The future to monitor for channel closure.  When this future
+            completes, the channel will be closed.
 
-    async def __pop(self) -> T:
+        Raises:
+            ValueError: If the provided future is None.
         """
-        Asynchronously retrieves a value from the queue.
+        if future is None:
+            raise ValueError('Cannot set a None future')
+
+        self._close_future = asyncio.ensure_future(future)
+        if self._close_future is not None:
```

Comment:
why check for None here? you made sure incoming future is not None already, this can never be None